services/ui_backend_service/data/cache/get_log_file_action.py

```diff
@@ -192,6 +202,147 @@ def log_size_exceeded_blurb(task: Task, logtype: str, max_size: int):
     return blurb
 
 
+def fetch_logs(task: Task, log_hash: str, logtype: str, force_reload: bool = False):
+    # TODO: This could theoretically be a part of the Metaflow client instead.
+    paths = []
+    stream = 'stderr' if logtype == STDERR else 'stdout'
+    log_location = task.metadata_dict.get('log_location_%s' % stream)
+
+    # TODO: Check that this is also under GC
+    log_tmp_path = os.path.join("cache_data", "log", "blobs", log_hash)
+    os.makedirs(log_tmp_path, exist_ok=True)
+    log_paths = {}
+
+    filecache = FileCache()
+    if log_location:
+        # Legacy log case
+        log_info = json.loads(log_location)
+        location = log_info["location"]
+        ds_type = log_info["ds_type"]
+        attempt = log_info["attempt"]
+        ds_cls = filecache._get_datastore_storage_impl(ds_type)
+        ds_root = ds_cls.path_join(*ds_cls.path_split(location)[:-5])
+
+        ds = filecache._get_task_datastore(
+            ds_type,
+            ds_root,
+            *task.path_components,
+            attempt
+        )
+        name = ds._metadata_name_for_attempt("%s.log" % stream)
+        to_load = [ds._storage_impl.path_join(ds._path, name)]
+        log_paths[name] = os.path.join(log_tmp_path, name)
+    else:
+        # MFLog support
+        meta_dict = task.metadata_dict
+        ds_type = meta_dict.get("ds-type")
+        ds_root = meta_dict.get("ds-root")
+        if ds_type is None or ds_root is None:
+            return
+
+        attempt = task.current_attempt
+
+        ds = filecache._get_task_datastore(
+            ds_type,
+            ds_root,
+            *task.path_components,
+            attempt
+        )
+        paths = dict(
+            map(
+                lambda s: (
+                    ds._metadata_name_for_attempt(
+                        ds._get_log_location(s, stream),
+                        attempt_override=False,
+                    ),
+                    s,
+                ),
+                LOG_SOURCES,
+            )
+        )
+        to_load = []
+        for name in paths.keys():
+            path = ds._storage_impl.path_join(ds._path, name)
+            to_load.append(path)
+            log_paths[name] = os.path.join(log_tmp_path, name)
+
+    # skip downloading as all files are on disk.
+    skip_dl = not force_reload and all(os.path.exists(path) for path in log_paths.values())
+    if not skip_dl:
+        # Load the log files to disk
+        with ds._storage_impl.load_bytes(to_load) as load_results:
+            for key, path, meta in load_results:
+                name = ds._storage_impl.basename(key)
+                if path is None:
+                    # no log file existed in the location. Usually not an error.
+                    log_paths[name] = None
+                else:
+                    shutil.move(path, log_paths[name])
+    return [val for val in log_paths.values() if val is not None]
```

Comment:
I wonder if this whole method could just be simplified by using the task datastore's load_logs (well through the filecache if you want to although not 100% sure) with the addition of something like "write_to_disk_only". In other words, you seem to be doing a lot of the thing from there with the only difference being that you don't open the logs but just keep them on disk. I feel we could make that small change there (without modifying any of the storage layers but just keeping it at the task datastore level) and this should simplify this code a lot. Basically, your line 280 (or somethin akin to it) would go around line 984 of task_datastore.py (protected by a flag or something).

Reply:
yep, as a follow up will look into hoisting all the logic required by fetch_logs into the Metaflow client side, for something akin to `dump_logs(stream, destination_path)`